RecoTracker/LSTCore/src/alpaka/TrackCandidate.h

```diff
@@ -0,0 +1,590 @@
+      lowerModuleIndices = buf.lowerModuleIndices_buf.data();
+
+      centerX = buf.centerX_buf.data();
+      centerY = buf.centerY_buf.data();
+      radius = buf.radius_buf.data();
+    }
+  };
+
+  template <typename TDev>
+  struct TrackCandidatesBuffer {
+    Buf<TDev, short> trackCandidateType_buf;
+    Buf<TDev, unsigned int> directObjectIndices_buf;
+    Buf<TDev, unsigned int> objectIndices_buf;
+    Buf<TDev, unsigned int> nTrackCandidates_buf;
+    Buf<TDev, unsigned int> nTrackCandidatespT3_buf;
+    Buf<TDev, unsigned int> nTrackCandidatespT5_buf;
+    Buf<TDev, unsigned int> nTrackCandidatespLS_buf;
+    Buf<TDev, unsigned int> nTrackCandidatesT5_buf;
+
+    Buf<TDev, uint8_t> logicalLayers_buf;
+    Buf<TDev, unsigned int> hitIndices_buf;
+    Buf<TDev, int> pixelSeedIndex_buf;
+    Buf<TDev, uint16_t> lowerModuleIndices_buf;
+
+    Buf<TDev, FPX> centerX_buf;
+    Buf<TDev, FPX> centerY_buf;
+    Buf<TDev, FPX> radius_buf;
+
+    TrackCandidates data_;
+
+    template <typename TQueue, typename TDevAcc>
+    TrackCandidatesBuffer(unsigned int maxTrackCandidates, TDevAcc const& devAccIn, TQueue& queue)
+        : trackCandidateType_buf(allocBufWrapper<short>(devAccIn, maxTrackCandidates, queue)),
+          directObjectIndices_buf(allocBufWrapper<unsigned int>(devAccIn, maxTrackCandidates, queue)),
+          objectIndices_buf(allocBufWrapper<unsigned int>(devAccIn, 2 * maxTrackCandidates, queue)),
+          nTrackCandidates_buf(allocBufWrapper<unsigned int>(devAccIn, 1, queue)),
+          nTrackCandidatespT3_buf(allocBufWrapper<unsigned int>(devAccIn, 1, queue)),
+          nTrackCandidatespT5_buf(allocBufWrapper<unsigned int>(devAccIn, 1, queue)),
+          nTrackCandidatespLS_buf(allocBufWrapper<unsigned int>(devAccIn, 1, queue)),
+          nTrackCandidatesT5_buf(allocBufWrapper<unsigned int>(devAccIn, 1, queue)),
+          logicalLayers_buf(allocBufWrapper<uint8_t>(devAccIn, Params_pT5::kLayers * maxTrackCandidates, queue)),
+          hitIndices_buf(allocBufWrapper<unsigned int>(devAccIn, Params_pT5::kHits * maxTrackCandidates, queue)),
+          pixelSeedIndex_buf(allocBufWrapper<int>(devAccIn, maxTrackCandidates, queue)),
+          lowerModuleIndices_buf(allocBufWrapper<uint16_t>(devAccIn, Params_pT5::kLayers * maxTrackCandidates, queue)),
+          centerX_buf(allocBufWrapper<FPX>(devAccIn, maxTrackCandidates, queue)),
+          centerY_buf(allocBufWrapper<FPX>(devAccIn, maxTrackCandidates, queue)),
+          radius_buf(allocBufWrapper<FPX>(devAccIn, maxTrackCandidates, queue)) {
+      alpaka::memset(queue, nTrackCandidates_buf, 0u);
+      alpaka::memset(queue, nTrackCandidatesT5_buf, 0u);
+      alpaka::memset(queue, nTrackCandidatespT3_buf, 0u);
+      alpaka::memset(queue, nTrackCandidatespT5_buf, 0u);
+      alpaka::memset(queue, nTrackCandidatespLS_buf, 0u);
+      alpaka::memset(queue, logicalLayers_buf, 0u);
+      alpaka::memset(queue, lowerModuleIndices_buf, 0u);
```

Comment:
@makortel @fwyzard 
I'm trying to move TrackCandidate to SoAs.
After I instantiate a `PortableCollection<::lst::TrackCandidatesSoA>(maxTrackCandidates, queue)`, should I still be calling `alpaka::memset` on corresponding data views as it is in this code or is it better to write an initialization/reset kernel?

Reply:
hi @slava77, doing a single `memset` on the whole buffer should be faster.
If you need to do only on some of the columns, I think (but did not measure) that a kernel might be faster than multiple `memset` calls.

Reply:
the following compiles but seems odd (why do I need to keep the temporary?)
```C++
cands_ = new PortableCollection<::lst::TrackCandidatesSoA>(maxTrackCandidates, queue);
auto buf = trackCandidatesDC_->buffer();
alpaka::memset(queue, buf, 0u);
```

The `buffer()` method returns a value, while the method takes an lvalue. Is this a normal code or an indication that I'm doing something wrong?

Reply:
https://alpaka-group.github.io/alpaka/namespacealpaka.html#aa7b4d9f42d421578d38e8923da784b37
says that the 2nd argument can be a temporary.

Reply:
(actual line numbers may differ, this is from the old build log):
```
src/RecoTracker/LSTCore/src/alpaka/Event.dev.cc:901:13: note: in instantiation of function template specialization 'alpaka::memset<alpaka::BufUniformCudaHipRt<alpaka::ApiHipRt, std::byte, s\
td::integral_constant<unsigned long, 1>, unsigned int>, alpaka::uniform_cuda_hip::detail::QueueUniformCudaHipRt<alpaka::ApiHipRt, false>>' requested here
    alpaka::memset(queue, trackCandidatesDC_->buffer(), 0u);
            ^
/cvmfs/cms.cern.ch/el8_amd64_gcc12/external/alpaka/1.1.0-0a6641b4bfdf883c8da5b9e8620be504/include/alpaka/mem/buf/uniformCudaHip/Set.hpp:240:40: note: candidate function [with TExtent = alpa\
ka::Vec<std::integral_constant<unsigned long, 1>, unsigned int>, TView = alpaka::BufUniformCudaHipRt<alpaka::ApiHipRt, std::byte, std::integral_constant<unsigned long, 1>, unsigned int>] no\
t viable: expects an lvalue for 1st argument
            ALPAKA_FN_HOST static auto createTaskMemset(TView& view, std::uint8_t const& byte, TExtent const& extent)

```

Reply:
... that was partial, here is a fresh one:
```
/cvmfs/cms.cern.ch/el8_amd64_gcc12/external/alpaka/1.1.0-0a6641b4bfdf883c8da5b9e8620be504/include/alpaka/mem/view/Traits.hpp:221:16: error: no matching function for call to 'createTaskMemset'
        return trait::CreateTaskMemset<Dim<TView>, Dev<TView>>::createTaskMemset(
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/cvmfs/cms.cern.ch/el8_amd64_gcc12/external/alpaka/1.1.0-0a6641b4bfdf883c8da5b9e8620be504/include/alpaka/mem/view/Traits.hpp:247:24: note: in instantiation of function template specialization 'alpaka::createTaskMemset<alpaka::Vec<std::integral_constant<unsigned long, 1>, unsigned int>, alpaka::BufUniformCudaHipRt<alpaka::ApiHipRt, std::byte, std::integral_constant<unsigned long, 1>, unsigned int>>' requested here
        enqueue(queue, createTaskMemset(std::forward<TViewFwd>(view), byte, getExtents(view)));
                       ^
src/RecoTracker/LSTCore/src/alpaka/Event.dev.cc:912:13: note: in instantiation of function template specialization 'alpaka::memset<alpaka::BufUniformCudaHipRt<alpaka::ApiHipRt, std::byte, std::integral_constant<unsigned long, 1>, unsigned int>, alpaka::uniform_cuda_hip::detail::QueueUniformCudaHipRt<alpaka::ApiHipRt, false>>' requested here
    alpaka::memset(queue, trackCandidatesDC_->buffer(), 0u);
            ^
/cvmfs/cms.cern.ch/el8_amd64_gcc12/external/alpaka/1.1.0-0a6641b4bfdf883c8da5b9e8620be504/include/alpaka/mem/buf/uniformCudaHip/Set.hpp:240:40: note: candidate function [with TExtent = alpaka::Vec<std::integral_constant<unsigned long, 1>, unsigned int>, TView = alpaka::BufUniformCudaHipRt<alpaka::ApiHipRt, std::byte, std::integral_constant<unsigned long, 1>, unsigned int>] not viable: expects an lvalue for 1st argument
            ALPAKA_FN_HOST static auto createTaskMemset(TView& view, std::uint8_t const& byte, TExtent const& extent)
                                       ^
```

is it related to `createTaskMemset(TView& view` vs `alpaka::memset ( TQueue &  	queue, TViewFwd &&  view,` ?

Reply:
> is it related to `createTaskMemset(TView& view` vs `alpaka::memset ( TQueue & queue, TViewFwd && view,` ?

That indeed seems to be the case. Digging into Alpaka code, I see the generic `alpaka::memset()` and `alpaka::createTaskMemset()` (called by `memset()`) use the universal reference, but in the implementations the CUDA, HIP, and SYCL implementations use the lvalue-reference (whereas the CPU implementation seems to use the universal reference).

Ok, I found also an open issue about this exact question https://github.com/alpaka-group/alpaka/issues/2292 . It seems to me that for now (for non-CPU backends) one needs to hold the "temporary" buffer object, while API may improve in the future.

Reply:
#45861 adds a member function `zeroInitialise()` to the `Portable` objects and collections.